uniffi_bindgen/src/bindings/kotlin/templates/EnumTemplate.kt

```diff
@@ -1,19 +1,94 @@
-enum class {{ e.name()|class_name_kt }} {
-    {% for variant in e.variants() %}
-    {{ variant|enum_variant_kt }}{% if loop.last %};{% else %},{% endif %}
+{#
+// Kotlin's `enum class` constuct doesn't support variants with associated data,
+// but is a little nicer for consumers than its `sealed class` enum pattern.
+// So, we switch here, using `enum class` for enums with no associated data
+// and `sealed class` for the general case.
+#}
+
+{% if e.has_associated_data() %}
+
+sealed class {{ e.name()|class_name_kt }} {
+    {% for variant in e.variants() -%}
+    class {{ variant.name()|class_name_kt }}({% if variant.has_fields() %}
+        {% for field in variant.fields() -%}
+        val {{ field.name()|var_name_kt }}: {{ field.type_()|type_kt}}{% if loop.last %}{% else %}, {% endif %}
+        {% endfor -%}
+    {%- endif %}) : {{ e.name()|class_name_kt }}() {
+        override fun write(buf: RustBufferBuilder) {
+            buf.putInt({{ loop.index }})
+            {% for field in variant.fields() -%}
+            {{ "(this.{})"|format(field.name())|write_kt("buf", field.type_()) }}
+            {% endfor -%}
+        }
+        override fun equals(other: Any?) : Boolean =
+            if (other is {{ e.name()|class_name_kt }}.{{ variant.name()|class_name_kt }}) {
+                {% if variant.has_fields() -%}
+                {% for field in variant.fields() -%}
+                {{ field.name()|var_name_kt }} == other.{{ field.name()|var_name_kt }}{% if loop.last %}{% else %} && {% endif -%}
+                {% endfor -%}
+                {% else -%}
+                true
+                {%- endif %}
+            } else {
+                false
+            }
+    }
     {% endfor %}
 
     companion object {
-        internal fun lift(n: Int) =
-            try { values()[n - 1] }
+        internal fun lift(rbuf: RustBuffer.ByValue): {{ e.name()|class_name_kt }} {
+            return liftFromRustBuffer(rbuf) { buf -> {{ e.name()|class_name_kt }}.read(buf) }
+        }
+
+        internal fun read(buf: ByteBuffer): {{ e.name()|class_name_kt }} {
+            return when(buf.getInt()) {
+                {%- for variant in e.variants() %}
+                {{ loop.index }} -> {{ e.name()|class_name_kt }}.{{ variant.name()|class_name_kt }}({% if variant.has_fields() %}
+                    {% for field in variant.fields() -%}
+                    {{ "buf"|read_kt(field.type_()) }}{% if loop.last %}{% else %},{% endif %}
+                    {% endfor -%}
+                {%- endif -%})
+                {%- endfor %}
+                else -> throw RuntimeException("invalid enum value, something is very wrong!!")
+            }
+        }
+    }
+
+    internal fun lower(): RustBuffer.ByValue {
+        return lowerIntoRustBuffer(this, {v, buf -> v.write(buf)})
+    }
+
+    internal open fun write(buf: RustBufferBuilder) {
+        throw RuntimeException("enum variant should have overridden `write` method, something is very wrong!!")
```

Comment:
On reflection, I think I'd rather put the `write` method and friends in the top level, and use pattern matching to work out what to do. 

(FWIW I don't think there's any perf difference, in fact: given it's thee idiomatic way of doing tagged-unions, I expect there's some particular optimizations we could end up losing out on by adding methods to individual variant classes.)